iopath/common/azure_blob.py

```diff
@@ -0,0 +1,632 @@
+            self._get_local_path_single(blob_path, force=force)
+
+        return result_path
+
+    def _get_local_path_single(self, path: str, force: bool = False) -> str:
+        local_path = self._local_cache_path(path)
+        with file_lock(local_path):
+            if os.path.exists(local_path):
+                # Redownload if remote is newer.
+                props = self._get_blob_properties(path)
+                remote_modified = props["last_modified"]
+                local_modified = datetime.fromtimestamp(
+                    os.path.getmtime(local_path)
+                ).astimezone()
+
+                if remote_modified <= local_modified and not force:
+                    logger.info(
+                        "URL {} was already cached in {}".format(path, local_path)
+                    )
+                    return local_path
+
+            logger.info("Caching {} ...".format(path))
+            tmp = local_path + ".tmp"
+            # clean-up tmp if found, because if tmp exists, it must be a dirty
+            # result of a previously process that didn't cleanup itself.
+            if os.path.isfile(tmp):
+                os.unlink(tmp)
+
+            account, container, blob_path = self._parse_uri(path)
+            client = self._get_client(account)
+            try:
+                blob_stream = client.get_blob_client(
+                    container=container, blob=blob_path
+                ).download_blob()
+
+                with open(tmp, "wb") as f_tmp:
+                    blob_stream.readinto(f_tmp)
+
+                shutil.move(tmp, local_path)
+            finally:
+                try:
+                    os.unlink(tmp)
+                except Exception:
+                    pass
+
+            logger.info("URL {} cached in {}".format(path, local_path))
+            return local_path
+
+    def _open(
+        self,
+        path: str,
+        mode: str = "rb",
+        buffering: int = -1,
+        **kwargs: Any,
+    ) -> Union[IO[str], IO[bytes]]:
+        """
+        Open a stream to a URI, similar to the built-in `open`.
+        Args:
+            path (str): A URI supported by this PathHandler
+            mode (str): Specifies the mode in which the file is opened. It defaults
+                to 'r'.
+            buffering (int): An optional integer used to set the buffering policy.
+                Pass 0 to switch buffering off and an integer >= 1 to indicate the
+                size in bytes of a fixed-size chunk buffer. When no buffering
+                argument is given, the default buffering policy depends on the
+                underlying I/O implementation.
+        Returns:
+            file: a file-like object.
+        """
+        self._check_kwargs(kwargs)
+        assert mode in ("rb", "wb"), "Supported modes: rb, wb"
+
+        chunk_size = buffering if buffering > 0 else DEFAULT_CHUNK_SIZE
+
+        account, container, blob_path = self._parse_uri(path)
```

Comment:
What happens if `blob_path` is path to a directory? Do you need a check to raise an exception in that case?

Reply:
Blob storage sort of simulates having a directory structure, but at the end of the day it's really just a flat hierarchy under a given storage container. I just tested this out and I get the following error which I think is descriptive enough. Let me know what you think:

```
azure.core.exceptions.ResourceNotFoundError: The specified blob does not exist.
RequestId:007bbeb1-001e-009f-58bb-fe96f3000000
Time:2022-11-22T21:47:16.5583532Z
ErrorCode:BlobNotFound
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
RequestId:007bbeb1-001e-009f-58bb-fe96f3000000
Time:2022-11-22T21:47:16.5583532Z</Message></Error>
```